tests/integration/test_kms.py

```diff
@@ -384,53 +387,111 @@ def test_generate_data_key_without_plaintext(self, kms_client, kms_key):
-    def test_sign_verify(self, kms_client, key_type, kms_create_key):
-        key_spec = "RSA_2048" if key_type == "rsa" else "ECC_NIST_P256"
-        result = kms_create_key(KeyUsage="SIGN_VERIFY", KeySpec=key_spec)
-        key_id = result["KeyId"]
-        # The Message parameter for Sign contains either a message itself, or a signature of a message. It is
-        # MessageType parameter that specifies one of the two cases. Unfortunately, at some point our code was
-        # ignoring the MessageType setting, so both types of message were getting treated the same way. Want to have
-        # a test to make sure that is no longer the case.
-        #
-        # The following string is 32 bytes, the same length as SHA256 digests we use for message signatures. So KMS
-        # accepts the string both as a plaintext message and a signature. As a result - we can use the same string as
-        # Message with different settings for MessageType to see if the outcome is different.
+    @pytest.mark.parametrize(
+        "key_spec,sign_algo",
+        [
+            ("RSA_2048", "RSASSA_PSS_SHA_256"),
+            ("RSA_2048", "RSASSA_PSS_SHA_384"),
+            ("RSA_2048", "RSASSA_PSS_SHA_512"),
+            ("RSA_4096", "RSASSA_PKCS1_V1_5_SHA_256"),
+            ("RSA_4096", "RSASSA_PKCS1_V1_5_SHA_512"),
+            ("ECC_NIST_P256", "ECDSA_SHA_256"),
+            ("ECC_NIST_P256", "ECDSA_SHA_384"),
+            ("ECC_SECG_P256K1", "ECDSA_SHA_256"),
+            ("ECC_SECG_P256K1", "ECDSA_SHA_512"),
+        ],
+    )
+    def test_sign_verify(self, kms_client, kms_create_key, key_spec, sign_algo):
+        hash_algo = get_hash_algorithm(sign_algo)
+        hasher = getattr(hashlib, hash_algo.replace("_", "").lower())
+
+        plaintext = b"test message !%$@ 1234567890"
+        digest = hasher(plaintext).digest()
+
+        key_id = kms_create_key(KeyUsage="SIGN_VERIFY", KeySpec=key_spec)["KeyId"]
+
+        kwargs = {"KeyId": key_id, "SigningAlgorithm": sign_algo}
+
+        bad_signature = kms_client.sign(MessageType="RAW", Message="bad", **kwargs)["Signature"]
+        bad_message = b"bad message 321"
+
+        # Ensure raw messages can be signed and verified
+        signature = kms_client.sign(MessageType="RAW", Message=plaintext, **kwargs)["Signature"]
+        assert kms_client.verify(
+            MessageType="RAW", Signature=signature, Message=plaintext, **kwargs
+        )["SignatureValid"]
+
+        # Ensure pre-hashed messages can be signed and verified
+        signature = kms_client.sign(MessageType="DIGEST", Message=digest, **kwargs)["Signature"]
+        assert kms_client.verify(
+            MessageType="DIGEST", Signature=signature, Message=digest, **kwargs
+        )["SignatureValid"]
+
+        # Ensure bad digest raises during signing
+        with pytest.raises(ClientError) as exc:
+            kms_client.sign(MessageType="DIGEST", Message=plaintext, **kwargs)
+        assert exc.match("ValidationException")
```

Comment:
nit: Since the tests are already AWS validated, we could also do a `snapshot.match(..)` here, which would also assert the other elements of the error response. (no need to change now, though, we could tackle the snapshots in a follow-up iteration 👍 ).

Reply:
Fixed :+1: